### ISSUE TITLE
feature idea: toggle fraction

### ISSUE THREAD

**Issue by nilsbecker** [lervag/vimtex#1390]
a feature that would ease editing a lot for me is the following. an automatic conversion between inline fractions `$a/b$` and display fractions `$\frac{a}{b}$`. ideally, this would be a command similar to `tsd` now, maybe `tf`? 

the direction from display to inline is relatively easy:
- invocation should work anywhere on the fraction command.
- first argument
- if it is not surrounded by parentheses and consists of more than one word, surround by parentheses
- "/"
- second argument
- same parentheses treatment as for first argument.
here the question is what a 'word' is. i think a simple definition like `W` in vim would be already useful.

the direction from inline to display involves deciding from where to invoke. 
- if there is a marked region, split that on '/' and make the first part the first argument and the second part the second argument. 
- if there is no marked region, the command can only work either when the cursor is on the numerator or on the denominator. a choice has to be made. let's say the denominator. then: search backwards to the '/'. search forward to 'E', the end of the current 'W' word. this span is the second argument. from the '/' search further backwards to the 'B' of the previous 'W'. 
- remove surrounding parentheses in the arguments if present
- output the `\fraction` command

does that sound within scope?

**Comment by kiryph** [comment]
This could also be something for the plugin https://github.com/AndrewRadev/switch.vim

Following is untested but should be close:
```viml
autocmd FileType tex let b:switch_custom_definitions =
      \ [
      \   {
      \     '$\(\S*\)/\(\S*\)$':        '$\frac{\1}{\2}$',
      \     ' \(\S*\)/\(\S*\) ':        ' \frac{\1}{\2} ',
      \   }
      \ ]
```

**Comment by nilsbecker** [comment]
interesting plugin! thanks for pointing it out. i'm unsure where this functionality should go, vimtex or another plugin

**Comment by nilsbecker** [comment]
i tried it out, now i have this:
```
autocmd FileType tex let b:switch_custom_definitions =
      \ [
      \   {
      \     '\(\S*\)/\(\S*\)' : '\\frac{\1}{\2}',
      \     '\\frac{\(\S*\)}{\(\S*\)}': '\1/\2' 
      \   }
      \ ]
```
which does not have the parentheses logic yet; also space on either side of the `/` is not recognized. so far so good.

**Comment by nilsbecker** [comment]
... but it does not respect math environment delimiters if they are adjacent without space. i guess the definition of what a 'word' is would have to be refined considerably.

**Comment by lervag** [comment]
It seems an interesting challenge, and I can see the usefulness of this command. I agree it could be out of scope, but at the same time, it seems a little bit overkill to install a new plugin for a single useful feature.

I feel it should not be too hard to implement this, and if you agree it may be useful as a new feature for vimtex then I will try to add it. What do you think about the following list of examples? Are there any cases missing?

```tex
$\frac x  {x-1}$     -> $x/(x-1)$
$\frac {x-1} x$      -> $(x-1)/x$
$\frac{x+1}{x-1}$    -> $(x+1)/(x-1)$
$\frac {x+1}  {x-1}$ -> $(x+1)/(x-1)$
$x / (x-1)$          -> $\frac x {x-1}$
$(x-1) /x$           -> $\frac {x-1} x$
$(x+1)  /(x-1)$      -> $\frac{x+1}{x-1}$
$(x+1)/ (x-1)$       -> $\frac{x+1}{x-1}$
```

**Comment by lervag** [comment]
Note: If I implement this, I will first make a `<plug>(vimtex-toggle-frac)` mapping that you can test and play around with. Only after we are happy with how things work will I consider to add a default mapping for it, and I think we should consider that carefully. `tf` is not really very good, since `t` is a **very** useful intrinsic mapping!

**Comment by nilsbecker** [comment]
cool. i tend to agree that installing a plugin for a single use case is not ideal. i would certainly try out the command for testing. i agree with your examples. one thing that is not contained is the extent of the numerator/denominator. it can't be automatically the whole math environment since one will often want just to convert part of a formula. so `$a b c / (d+e)$ -> $a b \frac{c}{d+e}$` or `-> $\frac{a b c}{d+e}$` can both be desired. this would have to be decided either based on cursor position but in general will probably require visual highlighting. (i.e. there should be a reasonable default for what to take as the extent of the fraction but if the user wants something else they have to highlight)

**Comment by lervag** [comment]
Discussion moved to the new PR, #1625.
